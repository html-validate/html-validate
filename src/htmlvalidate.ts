import { Config, ConfigLoader } from './config';
import { Engine, TokenDump, EventDump } from './engine';
import { Parser } from './parser';
import { Report } from './reporter';
import { Source } from './context';

class HtmlValidate {
	private globalConfig: Config;

	constructor(options?: any){
		const defaults = Config.empty();
		this.globalConfig = defaults.merge(options ? Config.fromObject(options) : Config.defaultConfig());
	}

	/**
	 * Parse HTML from string.
	 *
	 * @param str {string} - Text to parse.
	 * @return {object} - Report output.
	 */
	public validateString(str: string): Report {
		const source = {
			data: str,
			filename: 'inline',
			line: 1,
			column: 1,
		};
		return this.validateSource(source);
	}

	/**
	 * Parse HTML from source.
	 *
	 * @param str {string} - Text to parse.
	 * @return {object} - Report output.
	 */
	public validateSource(source: Source): Report {
		const config = this.getConfigFor('inline');
		const engine = new Engine(config, Parser);
		return engine.lint([source]);
	}

	/**
	 * Parse HTML from file.
	 *
	 * @param filename {string} - Filename to read and parse.
	 * @return {object} - Report output.
	 */
	public validateFile(filename: string): Report {
		const config = this.getConfigFor(filename);
		const source = config.transform(filename);
		const engine = new Engine(config, Parser);
		return engine.lint(source);
	}

	public dumpTokens(filename: string): TokenDump[] {
		const config = this.getConfigFor(filename);
		const source = config.transform(filename);
		const engine = new Engine(config, Parser);
		return engine.dumpTokens(source);
	}

	public dumpEvents(filename: string): EventDump[] {
		const config = this.getConfigFor(filename);
		const source = config.transform(filename);
		const engine = new Engine(config, Parser);
		return engine.dumpEvents(source);
	}

	public dumpTree(filename: string): string[] {
		const config = this.getConfigFor(filename);
		const source = config.transform(filename);
		const engine = new Engine(config, Parser);
		return engine.dumpTree(source);
	}

	/**
	 * Get parser for given filename.
	 */
	public getParserFor(source: Source): Parser {
		const config = this.getConfigFor(source.filename);
		return new Parser(config);
	}

	/**
	 * Get configuration for given filename.
	 */
	public getConfigFor(filename: string): Config {
		const loader = new ConfigLoader();
		const config = loader.fromTarget(filename);
		const merged = this.globalConfig.merge(config);
		merged.init();
		return merged;
	}
}

export default HtmlValidate;
