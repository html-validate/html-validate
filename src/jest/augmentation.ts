/** module augmentation: jest.d.ts */
import { type ConfigData } from "../config";
import { type Message } from "../message";

declare global {
	/* eslint-disable-next-line @typescript-eslint/no-namespace -- module augmentation */
	namespace jest {
		/* eslint-disable-next-line @typescript-eslint/no-empty-object-type, @typescript-eslint/no-unused-vars -- to match jest declaration */
		interface Matchers<R, T = {}> {
			toBeValid(): R;
			toBeInvalid(): R;
			toHaveError(error: Partial<Message>): R;
			toHaveError(ruleId: string, message: string, context?: any): R;
			toHaveErrors(errors: Array<[string, string] | Record<string, unknown>>): R;

			/**
			 * Validate string or HTMLElement.
			 *
			 * Test passes if result is valid.
			 *
			 * @param config - Optional HTML-Validate configuration object.
			 * @param filename - Optional filename used when matching transformer and
			 * loading configuration.
			 */
			toHTMLValidate(filename?: string): R;
			toHTMLValidate(config: ConfigData, filename?: string): R;
			toHTMLValidate(error: Partial<Message>, filename?: string): R;
			toHTMLValidate(error: Partial<Message>, config: ConfigData, filename?: string): R;

			/**
			 * Writes out the given [[Report]] using codeframe formatter and compares
			 * with snapshot.
			 */
			toMatchCodeframe(snapshot?: string): R;

			/**
			 * Writes out the given [[Report]] using codeframe formatter and compares
			 * with inline snapshot.
			 */
			toMatchInlineCodeframe(snapshot?: string): R;
		}
	}
}
/** module augmentation end */
