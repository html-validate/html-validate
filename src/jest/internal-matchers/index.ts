export { toBeToken } from "./to-be-token";
