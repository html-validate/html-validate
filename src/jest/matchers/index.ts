export { toBeValid } from "./to-be-valid";
export { toBeInvalid } from "./to-be-invalid";
export { toHTMLValidate } from "./to-htmlvalidate";
export { toHaveError } from "./to-have-error";
export { toHaveErrors } from "./to-have-errors";
export { toMatchCodeframe } from "./to-match-codeframe";
export { toMatchInlineCodeframe } from "./to-match-inline-codeframe";
