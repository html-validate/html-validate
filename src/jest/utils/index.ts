export { diff } from "./diff";
export { type MaybeAsyncCallback, diverge } from "./diverge";
export { flattenMessages } from "./flatten-messages";
export { isThenable } from "./is-thenable";
export { type MatcherContext } from "./matcher-context";
export { type MatcherExpect } from "./matcher-expect";
export { type MatcherResult } from "./matcher-result";
export { stripAnsi } from "./strip-ansi";
