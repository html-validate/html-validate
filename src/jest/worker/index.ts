export { type ValidateStringFn } from "./worker";
export { createSyncFn } from "./create-sync-fn";
export { default as workerPath } from "./worker.ts?worker&url";
