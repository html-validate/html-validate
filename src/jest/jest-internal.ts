import "./jest";

import { Location } from "../context";
import { TokenType } from "../lexer";
import { toBeToken } from "./internal-matchers";

interface TokenMatcher {
	type: TokenType;
	location?: Partial<Location>;
	data?: any;
}

declare global {
	/* eslint-disable-next-line @typescript-eslint/no-namespace */
	namespace jest {
		/* eslint-disable-next-line @typescript-eslint/ban-types, @typescript-eslint/no-unused-vars */
		interface Matchers<R, T = {}> {
			toBeToken(expected: TokenMatcher): R;
		}
	}
}

expect.extend({
	toBeToken,
});
