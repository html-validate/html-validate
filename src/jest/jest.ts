import "./augmentation";

import {
	toBeValid,
	toBeInvalid,
	toHTMLValidate,
	toHaveError,
	toHaveErrors,
	toMatchCodeframe,
	toMatchInlineCodeframe,
} from "./matchers";
import { diff } from "./utils";

expect.extend({
	toBeValid: toBeValid(),
	toBeInvalid: toBeInvalid(),
	toHTMLValidate: toHTMLValidate(expect, diff),
	toHaveError: toHaveError(expect, diff),
	toHaveErrors: toHaveErrors(expect, diff),
	toMatchCodeframe: toMatchCodeframe(),
	toMatchInlineCodeframe: toMatchInlineCodeframe(),
});
