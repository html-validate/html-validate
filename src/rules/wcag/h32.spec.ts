import { HtmlValidate } from "../../htmlvalidate";
import "../../jest";

describe("wcag/h32", () => {
	let htmlvalidate: HtmlValidate;

	beforeAll(() => {
		htmlvalidate = new HtmlValidate({
			root: true,
			elements: [
				"html5",
				{
					"my-form": {
						form: true,
					},
				},
			],
			rules: { "wcag/h32": "error" },
		});
	});

	it("should not report when form has nested submit button (button)", async () => {
		expect.assertions(1);
		const markup = '<form><button type="submit"></button></form>';
		const report = await htmlvalidate.validateString(markup);
		expect(report).toBeValid();
	});

	it("should not report when form has nested submit button (input)", async () => {
		expect.assertions(1);
		const markup = '<form><input type="submit"></form>';
		const report = await htmlvalidate.validateString(markup);
		expect(report).toBeValid();
	});

	it("should not report when form has nested submit button (image)", async () => {
		expect.assertions(1);
		const markup = '<form><input type="image"></form>';
		const report = await htmlvalidate.validateString(markup);
		expect(report).toBeValid();
	});

	it("should not report when form has associated submit button", async () => {
		expect.assertions(1);
		const markup = '<form id="foo"></form><button form="foo" type="submit"></button>';
		const report = await htmlvalidate.validateString(markup);
		expect(report).toBeValid();
	});

	it("should not report when form has both nested and associated submit button", async () => {
		expect.assertions(1);
		const markup = '<form id="foo"><button form="foo" type="submit"></button></form>';
		const report = await htmlvalidate.validateString(markup);
		expect(report).toBeValid();
	});

	it("should report error when form is missing submit button", async () => {
		expect.assertions(2);
		const markup = "<form></form>";
		const report = await htmlvalidate.validateString(markup);
		expect(report).toBeInvalid();
		expect(report).toMatchInlineCodeframe(`
			"error: <form> element must have a submit button (wcag/h32) at inline:1:2:
			> 1 | <form></form>
			    |  ^^^^
			Selector: form"
		`);
	});

	it("should report error when form only has regular button", async () => {
		expect.assertions(2);
		const markup = '<form><button type="button"></button></form>';
		const report = await htmlvalidate.validateString(markup);
		expect(report).toBeInvalid();
		expect(report).toMatchInlineCodeframe(`
			"error: <form> element must have a submit button (wcag/h32) at inline:1:2:
			> 1 | <form><button type="button"></button></form>
			    |  ^^^^
			Selector: form"
		`);
	});

	it("should report error when form is associated with regular button", async () => {
		expect.assertions(2);
		const markup = '<form id="foo"></form><button form="foo" type="button"></button>';
		const report = await htmlvalidate.validateString(markup);
		expect(report).toBeInvalid();
		expect(report).toMatchInlineCodeframe(`
			"error: <form> element must have a submit button (wcag/h32) at inline:1:2:
			> 1 | <form id="foo"></form><button form="foo" type="button"></button>
			    |  ^^^^
			Selector: #foo"
		`);
	});

	it("should report error when form nested button is associated with another form", async () => {
		expect.assertions(2);
		const markup = '<form id="foo"><button form="bar" type="submit"></button></form>';
		const report = await htmlvalidate.validateString(markup);
		expect(report).toBeInvalid();
		expect(report).toMatchInlineCodeframe(`
			"error: <form> element must have a submit button (wcag/h32) at inline:1:2:
			> 1 | <form id="foo"><button form="bar" type="submit"></button></form>
			    |  ^^^^
			Selector: #foo"
		`);
	});

	it("should support custom elements", async () => {
		expect.assertions(2);
		const markup = "<my-form></my-form>";
		const report = await htmlvalidate.validateString(markup);
		expect(report).toBeInvalid();
		expect(report).toMatchInlineCodeframe(`
			"error: <my-form> element must have a submit button (wcag/h32) at inline:1:2:
			> 1 | <my-form></my-form>
			    |  ^^^^^^^
			Selector: my-form"
		`);
	});

	it("smoketest", async () => {
		expect.assertions(1);
		const report = await htmlvalidate.validateFile("test-files/rules/wcag/h32.html");
		expect(report).toMatchInlineCodeframe(`
			"error: <form> element must have a submit button (wcag/h32) at test-files/rules/wcag/h32.html:9:2:
			   7 | </form>
			   8 |
			>  9 | <form>
			     |  ^^^^
			  10 | 	<input type="text">
			  11 | 	<button>Foo</button>
			  12 | 	<button type="button">Bar</button>
			Selector: form:nth-child(3)"
		`);
	});

	it("should contain documentation", async () => {
		expect.assertions(1);
		htmlvalidate = new HtmlValidate({
			rules: { "wcag/h32": "error" },
		});
		const docs = await htmlvalidate.getRuleDocumentation("wcag/h32");
		expect(docs).toMatchSnapshot();
	});
});
