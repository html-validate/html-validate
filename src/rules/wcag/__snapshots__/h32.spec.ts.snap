// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`wcag/h32 should contain documentation 1`] = `
{
  "description": "WCAG 2.1 requires each \`<form>\` element to have at least one submit button.",
  "url": "https://html-validate.org/rules/wcag/h32.html",
}
`;

exports[`wcag/h32 smoketest 1`] = `
[
  {
    "errorCount": 1,
    "filePath": "test-files/rules/wcag/h32.html",
    "messages": [
      {
        "column": 2,
        "context": undefined,
        "line": 9,
        "message": "<form> element must have a submit button",
        "offset": 110,
        "ruleId": "wcag/h32",
        "ruleUrl": "https://html-validate.org/rules/wcag/h32.html",
        "selector": "form:nth-child(3)",
        "severity": 2,
        "size": 4,
      },
    ],
    "source": "<form>
	<button type="submit">Submit</button>
</form>

<form>
	<input type="submit" value="Submit">
</form>

<form>
	<input type="text">
	<button>Foo</button>
	<button type="button">Bar</button>
</form>
",
    "warningCount": 0,
  },
]
`;
